### PR TITLE
8286270: [java.desktop] Replace color search in XColors with a switch statement

### PR DESCRIPTION
**Pull request by SWinxy**
The current implementation creates ~750 new objects on load, and uses a binary array search with a Comparable implementation. This implementation is probably bad. Instead, we can use the enhanced switch feature to possibly save time, resources, and code size. The new implementation means that the array doesn't have to be sorted alphanumerically necessarily, and it creates the colors on demand.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286270](https://bugs.openjdk.org/browse/JDK-8286270): [java.desktop] Replace color search in XColors with a switch statement

### Reviewers
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**) ⚠️ Review applies to [32640e43](https://git.openjdk.org/jdk/pull/7096/files/32640e43df2b33fd28d074dcb978682e8b7b4fb7)
 * [Alexander Zvegintsev](https://openjdk.org/census#azvegint) (@azvegint - **Reviewer**) ⚠️ Review applies to [32640e43](https://git.openjdk.org/jdk/pull/7096/files/32640e43df2b33fd28d074dcb978682e8b7b4fb7)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/7096/head:pull/7096` \
`$ git checkout pull/7096`

Update a local copy of the PR: \
`$ git checkout pull/7096` \
`$ git pull https://git.openjdk.org/jdk pull/7096/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7096`

View PR using the GUI difftool: \
`$ git pr show -t 7096`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/7096.diff">https://git.openjdk.org/jdk/pull/7096.diff</a>

</details>

### CODE REVIEW

**Comment by SWinxy** [comment]
Keeping this alive; the PR is being blocked by the OCA.

**Comment by SWinxy** [comment]
Keeping open.

**Comment by TheShermanTanker** [comment]
Hey there @SWinxy, same thing with this PR, this time rename the title to just 8286269

File: src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/XColors.java [hunk 1]
```diff
@@ -25,805 +25,531 @@
...
-import java.util.Arrays;
 import javax.swing.plaf.ColorUIResource;
 
 /**
- * @author  Shannon Hickey
+ * Utility class to convert color names to colors.
+ *
+ * @author Shannon Hickey
+ * @author SWinxy
  */
```
Comment (prrace) [inline src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/XColors.java:35]: You aren't the author, you are changing the file, and if we go anywhere near @author tags it is to remove them, not add them. Plus "Swinxy" is not very professional

**Review by prrace — CHANGES_REQUESTED** [review]
> The current implementation creates ~750 new objects on load

Whereas yours creates an un-bounded number .. and changes that two lookups for the same name will return the same "==" object. 

If you can find a compatible way, that avoids both these problems . I can reconsider

**Comment by ExE-Boss** [comment]
> > The current implementation creates ~750 new objects on load
> 
> Whereas yours creates an un-bounded number .. and changes that two lookups for the same name will return the same "==" object.

That’s not what happens, since `XColors::lookupColor` always calls `new ColorUIResource(…)` in both the old and new version.

--------------------------------------------------------------------------------

The difference is that the old version would perform **O**(log <var>n</var>) lookups with **O**(<var>n</var>) string comparisons per lookup, whereas the new one performs a single `String.hashCode()` invocation (the result of which is cached[^1]) followed by an **O**(log <var>n</var>) `lookupswitch` and then an **O**(1) `tableswitch`.

[^1]: https://github.com/openjdk/jdk/blob/9583e3657e43cc1c6f2101a64534564db2a9bd84/src/java.base/share/classes/java/lang/String.java#L2335-L2355

**Comment by prrace** [comment]
> That’s not what happens, since XColors::lookupColor always calls new ColorUIResource(…) in both the old and new version.

You are correct. I over-looked the toColor() implementation.

Ok. this should be fine once the author tag is removed.

File: src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/XColors.java (hunk 1 above)
Comment (SWinxy) [reply src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/XColors.java:32]: Done. I wasn't aware of the intent to get rid of the `@author` tags, but it's very valid. They've been removed.

**Review by prrace — APPROVED** [review]

**Comment by SWinxy** [comment]
Thank you @prrace! I will start communicating on the mailing list before making any changes from now on.

**Review by azvegint — CHANGES_REQUESTED** [review]

**Review by prrace — CHANGES_REQUESTED** [review]
good catch. These need to be fixed.

**Comment by SWinxy** [comment]
I did an automated test to see if those were the only other things wrong. Those were the only things that were wrong.

**Review by azvegint — APPROVED** [review]

**Review by prrace — APPROVED** [review]

**Comment by mrserb** [comment]
Do you have any numbers to compare possible performance improvements in lookupColor()? I'll assume it should be faster but can we prove that?

**Comment by prrace** [comment]
Hello @SWinxy this PR needs some attention

**Comment by prrace** [comment]
> I've hesitated asking for help, but the builds aren't working and I can't tell _why_.

the GHA builds are passing so far as I can tell.

**Comment by TheShermanTanker** [comment]
> I've hesitated asking for help, but the builds aren't working and I can't tell _why_.

Don't worry about it, sometimes the tests fail due to issues unrelated to the PR, you can check the tests that failed from within the PR itself

**Comment by SWinxy** [comment]
I can see which test suites are ran, sure. But they don't give me feedback as to what issues it had (good to know I got an error on the 4282 tests... but which *one*?). The bot removed the "needs sponsor" tag because tests had failed post-hoc (and they've failed for not being able to load stuff prior), so my PR surviving is dependent on those tests.

**Comment by SWinxy** [comment]
Awesome! Thanks lots for your help @prrace.

**Pull request closed without merging**